crates/agent/src/lib.rs

```diff
@@ -107,39 +111,53 @@ async fn listen_for_tasks(
     });
 
     loop {
-        // try_recv returns None when the channel disconnects,
-        // which we want to have explicit handling for
-        let maybe_item = listener
-            .lock()
-            .await
-            .try_recv()
-            .await
-            .map_err(|e| anyhow::Error::from(e))?;
-
-        if let Some(item) = maybe_item {
-            let notification: AgentNotification = serde_json::from_str(item.payload())
-                .context("deserializing agent task notification")?;
-
-            tracing::debug!(
-                table = &notification.table,
-                "Message received to invoke handler"
-            );
-            task_tx
-                .send(HandlerInvocation {
-                    table_name: notification.table,
-                    is_poll: false,
-                })
-                .map_err(|e| anyhow::Error::from(e))?
-        } else {
-            tracing::warn!("LISTEN/NOTIFY stream from postgres lost, waking all handlers and attempting to reconnect");
-            table_names.iter().for_each(|table| {
-                task_tx
-                    .send(HandlerInvocation {
-                        table_name: table.clone(),
-                        is_poll: false,
-                    })
-                    .unwrap()
-            });
+        // We need to release the lock on listener every once in a while so that the
+        // keepalive loop above can do its thing, otherwise we'll hold the lock open
+        // waiting on try_recv, which may never complete because the connection is
+        // unknowningly disconnected, which is the whole point of the keepalive loop.
+        match tokio::time::timeout(Duration::from_secs(30), async {
```

Comment:
I think Mutex is the wrong tool for the job here, and what you really want is `tokio::select!`.

You can have a single select where you're blocking for the first of _either_ a `NOTIFY` _or_ for a its-time-to-subscribe-and-unsubscribe-timer to elapse (upon which you'd synchronously subscribe / unsubscribe).

Then, remove the tokio::spawn loop above and `cloned_listener` altogether.

Reply:
Yuuuuuuup that does make more sense